sql/spanner/inspect.go

```diff
@@ -0,0 +1,738 @@
+	defer rows.Close()
+	for rows.Next() {
+		var tSchema, name, parentTable, onDeleteAction, spannerState sql.NullString
+		if err := rows.Scan(&tSchema, &name, &parentTable, &onDeleteAction, &spannerState); err != nil {
+			return fmt.Errorf("scan table information: %w", err)
+		}
+		if !sqlx.ValidString(name) {
+			return fmt.Errorf("invalid table name: %q", name.String)
+		}
+		sName := tSchema.String
+		if sName == "" {
+			sName = defaultSchemaNameAlias
+		}
+		s, ok := realm.Schema(sName)
+		if !ok {
+			return fmt.Errorf("schema %q was not found in realm", tSchema.String)
+		}
+		t := &schema.Table{
+			Name: name.String,
+		}
+		s.AddTables(t)
+		// TODO(tmc): handle parentTable, onDeleteAction, spannerState as attrs
+	}
+	if rows.Err() != nil {
+		return rows.Err()
+	}
+	return rows.Close()
+}
+
+// columns queries and appends the columns of the given table.
+func (i *inspect) columns(ctx context.Context, s *schema.Schema) error {
+	query := columnsQuery
+	rows, err := i.querySchema(ctx, query, s)
+	if err != nil {
+		return fmt.Errorf("spanner: querying schema %q columns: %w", s.Name, err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		if err := i.addColumn(s, rows); err != nil {
+			return fmt.Errorf("spanner: %w", err)
+		}
+	}
+	if err := rows.Err(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// addColumn scans the current row and adds a new column from it to the table.
+func (i *inspect) addColumn(s *schema.Schema, rows *sql.Rows) error {
+	var (
+		table, column, spannerType, colDefault sql.NullString
+		genExpr, spannerState                  sql.NullString
+		nullable, stored, generated            sql.NullBool
+		ord                                    sql.NullInt64
+
+		err error
+	)
+	if err := rows.Scan(
+		&table, &column, &ord, &colDefault,
+		&nullable, &spannerType, &generated, &genExpr,
+		&stored, &spannerState,
+	); err != nil {
+		return err
+	}
+	t, ok := s.Table(table.String)
+	if !ok {
+		return fmt.Errorf("table %q was not found in schema", table.String)
+	}
+	c := &schema.Column{
+		Name: column.String,
+		Type: &schema.ColumnType{
+			Raw:  spannerType.String,
+			Null: nullable.Bool,
+		},
+	}
+
+	// Converts spanner string type to schema.Type.
+	c.Type.Type, err = columnType(spannerType.String)
+	if err != nil {
+		return fmt.Errorf("spanner: Unable to convert string %q to schema.Type: %w", spannerType.String, err)
+	}
+
+	if colDefault.Valid {
+		c.Default = defaultExpr(c, colDefault.String)
+	}
+	if generated.Bool {
+		c.Attrs = []schema.Attr{
+			&schema.GeneratedExpr{
+				Expr: genExpr.String,
+				Type: "STORED",
+			},
+		}
+	}
+	t.AddColumns(c)
+	return nil
+}
+
+// Converts spanner string type to schema.Type.
+func columnType(spannerType string) (schema.Type, error) {
+	var typ schema.Type
+
+	col := &columnDesc{}
+	spannerType = strings.TrimSpace(strings.ToUpper(spannerType))
+
+	// Split up type into, base type, size, and other modifiers.
+	re := regexp.MustCompile(`(\w+)(?:\((-?\d+|MAX)\))?`)
```

Comment:
Move it outside this function, we can compile it once. e.g. https://github.com/ariga/atlas/blob/master/sql/postgres/inspect.go#L572